pkg/resource/deploy/plan.go

```diff
@@ -0,0 +1,605 @@
+package deploy
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/mitchellh/copystructure"
+
+	"github.com/pulumi/pulumi/pkg/v3/resource/deploy/providers"
+	"github.com/pulumi/pulumi/pkg/v3/version"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/config"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// A Plan is a mapping from URNs to ResourcePlans. The plan defines an expected set of resources and the expected
+// inputs and operations for each. The inputs and operations are treated as constraints, and may allow for inputs or
+// operations that do not exactly match those recorded in the plan. In the case of inputs, unknown values in the plan
+// accept any value (including no value) as valid. For operations, a same step is allowed in place of an update or
+// a replace step, and an update is allowed in place of a replace step. All resource options are required to match
+// exactly.
+type Plan struct {
+	ResourcePlans map[resource.URN]*ResourcePlan
+	Manifest      Manifest
+	// Any environment variables that were set when the plan was created. Values are encrypted.
+	EnvironmentVariables map[string]string
+	// The configuration in use during the plan.
+	Config config.Map
+}
+
+func NewPlan(config config.Map) Plan {
+	manifest := Manifest{
+		Time:    time.Now(),
+		Version: version.Version,
+		// Plugins: sm.plugins, - Explicitly dropped, since we don't use the plugin list in the manifest anymore.
+	}
+	manifest.Magic = manifest.NewMagic()
+
+	environmentVariables := make(map[string]string)
+	for _, e := range os.Environ() {
+		// golang has so many questionable design decisions, lets just hope no one puts '=' in there env var names.
+		pair := strings.SplitN(e, "=", 2)
+		environmentVariables[pair[0]] = pair[1]
+	}
+
+	return Plan{
+		ResourcePlans:        make(map[resource.URN]*ResourcePlan),
+		EnvironmentVariables: environmentVariables,
+		Manifest:             manifest,
+		Config:               config,
+	}
+}
+
+// Clone makes a deep copy of the given plan and returns a pointer to the clone.
+func (plan *Plan) Clone() *Plan {
+	return copystructure.Must(copystructure.Copy(plan)).(*Plan)
+}
+
+// PlanDiff holds the results of diffing two object property maps.
+type PlanDiff struct {
+	Adds    resource.PropertyMap   // the resource's properties we expect to add.
+	Deletes []resource.PropertyKey // the resource's properties we expect to delete.
+	Updates resource.PropertyMap   // the resource's properties we expect to update.
+}
```

Comment:
I'm a little surprised by the way we're storing diffs.

I think that using the inputs like this imposes a very strong requirement that all changes are visible via the inputs--i.e. there are no input vs. state differences. This is different from the model used by providers, where diffs can be observed either from input <-> input or from input <-> state. An alternative approach would be to store the detailed diff as returned by the provider (polyfilling a detailed diff where necessary), then checking that the detailed diff from the update is constrained to the detailed diff from the preview, with old values sourced from the old inputs or old state as directed by the detailed diff.

I think that what is implemented might be okay assuming that in the vast majority of cases all input <-> state differences are essentially input <-> input differences b/c the differing property has the same value in the state and the inputs (though that assumption requires that the provider's `Read` method can accurately reflect values from the state back to the inputs).

I was also a little surprised that we're storing a flat list of top-level diffs rather than the tree diff (ala `resource.ValueDiff`). Is there a benefit to the flat list? Storing the tree diff might make error presentation a little more precise.

Reply:
> I think that using the inputs like this imposes a very strong requirement that all changes are visible via the inputs

What would an input state difference be? I'm not sure I follow how we could have a state difference and not also have an input difference?

> Is there a benefit to the flat list?

Depends on what semantics we're going for. Like if you have a dict property and you run a plan to go from`{a: 1, b: 2}` to `{a: 1, c: 3}` is that recording "update to `{a: 1, c: 3}`" or "delete key b, add key c with 3"? Because the former only allows a plan to execute if it ends up with the value `{a: 1, c: 3}` but the later would allow someone to come and change the dict to something like `{b: 2, d: 4}` and the plan would allow the value `{c: 3, d: 4}` but not allow the value `{a: 1, c: 3}`.

Might be that second semantic is useful but it seemed more complex to code and explain and I don't have a concrete scenario that would use it so I err'ed on simplicity.

Reply:
> What would an input state difference be? I'm not sure I follow how we could have a state difference and not also have an input difference?

Consider a resource `R` with `(inputs, state) => ({foo: bar}, {foo: bar})`. If we refresh `R` and the provider returns `({foo: bar}, {foo: baz})`, the provider is still allowed to indicate that `R` needs to be updated on a subsequent run even if the program presents the inputs `{foo: bar}` b/c `foo` has different properties in the inputs and the state. This is arguably a bug, and I'm not sure it happens in practice, but it is a "feature" of the model that we might need to consider.

More common is the case where an input has no corresponding entry in the state, so diffs are only detectable from input to input (e.g. k8s secrets).

> the later would allow someone to come and change the dict to something like {b: 2, d: 4} and the plan would allow the value {c: 3, d: 4} but not allow the value {a: 1, c: 3}.

I think the latter would still allow `{a: 1, c: 3}` b/c the `same` on `c` is constrained to the `add` in the diff. I think the argument here is the same as for the analogous case at top level in that we want to allow the base state to differ as long as the diffs are compatible.

Reply:
This is good feedback and it sounds like we need to give it a thought. Would it be okay to go ahead with this PR to kick off examples testing and do this work as a follow-up?

Reply:
Yep, definitely. We can probably even ship the preview as-is and revisit this while we gather more data.